[SPARK-18148][SQL] Misleading Error Message for Aggregation Without Window/GroupBy

jiangxb1987 opened the pull request:
## What changes were proposed in this pull request?

Aggregation Without Window/GroupBy expressions will fail in `checkAnalysis`, the error message is a bit misleading, we should generate a more specific error message for this case.

For example,

```
spark.read.load("/some-data")
  .withColumn("date_dt", to_date($"date"))
  .withColumn("year", year($"date_dt"))
  .withColumn("week", weekofyear($"date_dt"))
  .withColumn("user_count", count($"userId"))
  .withColumn("daily_max_in_week", max($"user_count").over(weeklyWindow))
)
```

creates the following output:

```
org.apache.spark.sql.AnalysisException: expression '`randomColumn`' is neither present in the group by, nor is it an aggregate function. Add to group by or wrap in first() (or first_value) if you don't care which value you get.;
```

In the error message above, `randomColumn` doesn't appear in the query(acturally it's added by function `withColumn`), so the message is not enough for the user to address the problem.
## How was this patch tested?

Manually test

Before:

```
scala> spark.sql("select col, count(col) from tbl")
org.apache.spark.sql.AnalysisException: expression 'tbl.`col`' is neither present in the group by, nor is it an aggregate function. Add to group by or wrap in first() (or first_value) if you don't care which value you get.;;
```

After:

```
scala> spark.sql("select col, count(col) from tbl")
org.apache.spark.sql.AnalysisException: grouping expressions sequence is empty, and 'tbl.`col`' is not an aggregate function. Wrap '(count(col#231L) AS count(col)#239L)' in windowing function(s) or wrap 'tbl.`col`' in first() (or first_value) if you don't care which value you get.;;
```

Also add new test sqls in `group-by.sql`.

jiangxb1987 committed:
improve error message when checkAnalysis fail on Aggregate operator.

jiangxb1987 committed:
better format.

SparkQA commented:
**[Test build #67702 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67702/consoleFull)** for PR 15672 at commit [`350b7a3`](https://github.com/apache/spark/commit/350b7a335038adbe8a5a766d7a77fa17db0dbdeb).

SparkQA commented:
**[Test build #67702 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67702/consoleFull)** for PR 15672 at commit [`350b7a3`](https://github.com/apache/spark/commit/350b7a335038adbe8a5a766d7a77fa17db0dbdeb).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67702/
Test FAILed.

rxin commented:
Is this confusing msg reproducible using SQL queries (not dataframe APIs)? If yes, perhaps we can add a test in SQLQueryTestSuite.

jiangxb1987 committed:
add test cases.

jiangxb1987 commented:
@rxin I've added new test sqls in `group-by.sql`, which covers the both error message generated by `checkAnalysis`. I also moved the test cases for `Aggregate` operator from `SQLQuerySuite` to `group-by.sql`.

SparkQA commented:
**[Test build #67755 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67755/consoleFull)** for PR 15672 at commit [`8986207`](https://github.com/apache/spark/commit/8986207252e244cfadf9c5bc34264346b585bf18).

SparkQA commented:
**[Test build #67755 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67755/consoleFull)** for PR 15672 at commit [`8986207`](https://github.com/apache/spark/commit/8986207252e244cfadf9c5bc34264346b585bf18).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67755/
Test FAILed.

jiangxb1987 committed:
map aggregate expressions to sql

SparkQA commented:
**[Test build #67762 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67762/consoleFull)** for PR 15672 at commit [`033f43f`](https://github.com/apache/spark/commit/033f43fdf7835c24423c456dd7f26fe0c12a766a).

SparkQA commented:
**[Test build #67762 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67762/consoleFull)** for PR 15672 at commit [`033f43f`](https://github.com/apache/spark/commit/033f43fdf7835c24423c456dd7f26fe0c12a766a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67762/
Test FAILed.

jiangxb1987 commented:
The failed test case looks not related to our change here. It didn't fail in my local envirement.

jiangxb1987 commented:
retest this please

SparkQA commented:
**[Test build #67803 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67803/consoleFull)** for PR 15672 at commit [`033f43f`](https://github.com/apache/spark/commit/033f43fdf7835c24423c456dd7f26fe0c12a766a).

SparkQA commented:
**[Test build #67803 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67803/consoleFull)** for PR 15672 at commit [`033f43f`](https://github.com/apache/spark/commit/033f43fdf7835c24423c456dd7f26fe0c12a766a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67803/
Test PASSed.

rxin commented on sql/core/src/test/resources/sql-tests/inputs/group-by.sql:
```diff
@@ -15,3 +15,26 @@ select 'foo', approx_count_distinct(int_col) from myview where int_col == 0 grou
+SELECT a, COUNT(b) FROM testData;
+SELECT COUNT(a), COUNT(b) FROM testData;
+
+-- Aggregate with non-empty GroupBy expressions.
```
these are already tested earlier ain't they?

jiangxb1987 committed:
modify test cases.

SparkQA commented:
**[Test build #67888 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67888/consoleFull)** for PR 15672 at commit [`2542f7f`](https://github.com/apache/spark/commit/2542f7fcc9b19b0b577afb4a6bd6152909e0dd0c).

jiangxb1987 committed:
modify test cases.

jiangxb1987 commented:
@rxin I've updated test sqls to use one dataset. Thank you!

SparkQA commented:
**[Test build #67889 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67889/consoleFull)** for PR 15672 at commit [`43f8fa6`](https://github.com/apache/spark/commit/43f8fa6d642fb5eae3de13a5f88d592d6dcd4483).

SparkQA commented:
**[Test build #67889 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67889/consoleFull)** for PR 15672 at commit [`43f8fa6`](https://github.com/apache/spark/commit/43f8fa6d642fb5eae3de13a5f88d592d6dcd4483).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67889/
Test FAILed.

SparkQA commented:
**[Test build #67888 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67888/consoleFull)** for PR 15672 at commit [`2542f7f`](https://github.com/apache/spark/commit/2542f7fcc9b19b0b577afb4a6bd6152909e0dd0c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67888/
Test PASSed.

rxin commented:
Thanks - merging in master.
